### DISCUSSION TITLE
Matching error when there are no particles in a given frame

### DISCUSSION THREAD

**Discussion by stebrizzo** [ronshnapp/MyPTV#44] [Q&A]
Hi MyPTV community,

I noted a "bug" about matching module. When calling "matching" with the workflow file, the cycle stop when no particles are present or when no particles can be matched in a given frame (see attached screenshot for the error).

Where do you believe is the best place to fix this? Modifying the workflow file or the module itself?

<img width="1440" alt="matching_fail" src="https://github.com/ronshnapp/MyPTV/assets/46709347/d1bf2feb-6260-4e8a-bd47-df4f2daabfbd">

**Comment by stebrizzo** [comment]
I provvisionali solved the problem by using "try" in the "do_matching" section in the workflow file in this way, but it doesn't seem to work properly.

        if march_forwards==True:
            print('Matching forwards. Frames: %d -> %d'%(frames[0], frames[-1]))
            for f in frames:
 
                try:
                    mps.match_frame(f)
                except:
                    pass

**Comment by ronshnapp** [comment]
Hi Stefano,

Thanks for your comment! 

I actually know this bug but I was too lazy to fix it. But, with the help of your report, I finally tried to fix it, so I just pushed a small update to the master branch. Could you please try to download the new version, reinstall MyPTV, and let me know if this works?

In case it won't work, please share with me you camera and blob files, and then I'll be able to properly do the debugging.  

Best,
Ron

**Comment by alegambino** [comment]
Hi Ron and Stefano,

I have tried to download the new version of MyPTV and run the matching module on the same set of data Stefano was using. I found that my matching stops at the frame (frame 9, the same as in Stefano's error) where no particles have been segmented and my error is a bit different from Stefanos from what I see from the terminal.

![Screenshot 2024-02-21 102057](https://github.com/ronshnapp/MyPTV/assets/140545910/d8025ef7-e91d-4670-8d64-d4c2216c209a)

For completeness, I also post the data we've been using with segmented images (blob_cam files) and camera files of the setup.
[data.zip](https://github.com/ronshnapp/MyPTV/files/14357511/data.zip)

Thank you very much for this,

Best,

Alessandro

**Comment by ronshnapp** [comment]
Hi,

Sorry for the else instead of except - silly typo.
I fixed it on my PC and I think it should now work for you. Please download and reinstall again the updates on branch Master and let me know if it works. I also attach here the particles file that resulted from your data.
[particles.txt](https://github.com/ronshnapp/MyPTV/files/14365030/particles.txt)

Best,
Ron

**Comment by stebrizzo** [comment]
Hei Ron,

Just re-downloaded and re-installed MyPTV from the master branch on a new clear environemnt and tried matching with the same data alessandro provided, but the matching still fails at frame 9 (the one with no segmented particles) for some reasons:
<img width="1440" alt="Schermata 2024-02-22 alle 15 29 13" src="https://github.com/ronshnapp/MyPTV/assets/46709347/d030cd18-fb7d-4804-82cc-54ef2856c698">


By starting the matching from frame 10, it breaks at frame 90 with this error:
<img width="1440" alt="Schermata 2024-02-22 alle 15 32 20" src="https://github.com/ronshnapp/MyPTV/assets/46709347/6e979b53-cfad-4aa8-b5e7-852e0ff362ec">

I guess this happens when at least one of the blob files does not contain any blobs in a single frame (indeed cam 2 does not contain any blobs for frame later the 89).

What do you think?

**Comment by ronshnapp** [comment]
Hi Stefano,

It seems we probably are usin different versions of Scipy which leads to the error in my attempted workaround. 
I tried another solution cleaner solution. Can you try it out now?

Best,
Ron

**Comment by stebrizzo** [comment]
Hi Ron,
Great, it seems to work for this version =)

It is a bit strange that it was related to the scipy version, as I was using a clean virtual environment + installing the requirements directly from your files. Anyway it seems to work now, thanks!

Stefano

**Comment by ronshnapp** [comment]
Good. It is likely that I have an older scipy version installed.

Good luck with your experiments, and don't forget to share your cool results :)

Ron
